### ISSUE TITLE
[ROCm] GGPT inference runs successfully on MI300X + ROCm 7.2

### ISSUE THREAD

**Issue by ZJLi2013** [ChenYutongTHU/GGPT#7]
Hi, I successfully ran the full GGPT demo pipeline (VGGT → RoMaV2 → SfM → PTv3 refinement) on AMD MI300X with ROCm 7.2. Sharing the setup notes and one key gotcha (spconv weight layout) in case it helps other AMD users.

## Environment

| Component | Version |
|-----------|---------|
| GPU | AMD Instinct MI300X (gfx942) |
| Docker | `rocm/pytorch:rocm7.2_ubuntu24.04_py3.13_pytorch_release_2.10.0` |
| PyTorch | 2.10.0+rocm7.2.0 |
| spconv | [spconv_rocm](https://github.com/ZJLi2013/spconv_rocm) 2.3.8+rocm1 |
| flash-attn | [ROCm/flash-attention](https://github.com/ROCm/flash-attention) 2.8.4 (Triton backend) |

## ROCm Dependency Substitutions

| Original | ROCm Replacement | Notes |
|----------|-----------------|-------|
| `spconv-cu121` | `spconv_rocm` (pip install -e) | **Requires weight conversion** (see below) |
| `flash-attn` | `ROCm/flash-attention` (Triton backend) | Drop-in replacement |
| `torch-scatter` | Pure PyTorch shim (`segment_csr` via `scatter_add_`) | Only `segment_csr` is needed |
| `torch_geometric` | `pip install torch_geometric` (pure Python 2.7.0) | Works as-is |
| Pointcept `pointops` | Skipped (not called in demo) | Just add Pointcept to `PYTHONPATH` |

No hard blockers — all core dependencies have working ROCm alternatives.

## Key Fix: spconv Weight Layout Conversion

The `spconv_rocm` backend expects SubMConv3d weights in `[kernel_volume, in_ch, out_ch]` layout, while the released checkpoint uses the spconv v2 CUDA layout `[out_ch, kD, kH, kW, in_ch]`. Loading without conversion produces silent wrong results or shape errors.

The fix is a one-time permutation of the 22 CPE (Contextual Position Embedding) conv weights:

```python
import torch

ckpt = torch.load('ckpts/model.step228000.pth', map_location='cpu')
new = {}
for k, v in ckpt.items():
    if 'cpe' in k and 'weight' in k and v.dim() == 5:
        o, kd, kh, kw, i = v.shape
        new[k] = v.permute(1, 2, 3, 4, 0).reshape(kd*kh*kw, i, o)
    else:
        new[k] = v
torch.save(new, 'ckpts/model.step228000.rocm.pth')
```

Then update `configs/demo.yaml` to point to the converted checkpoint.

## Results

Pipeline timing on MI300X:

| Stage | Time |
|-------|------|
| VGGT feedforward | ~40s |
| RoMaV2 matching (7 pairs) | 4m05s |
| SfM BA + DLT | ~3s |
| GGPT PTv3 refinement (49 iter) | 5m44s |
| **Total** | **~11min** |

Output quality comparison (`scannetpp-9bb22982672c69bf`):

| Output | Points | Bbox Extent | Std(xyz) |
|--------|--------|-------------|----------|
| `ggpt_points.ply` | 1,000,000 | [2.72, 2.87, 3.51] | [0.572, 0.531, 0.616] |
| `ff_points.ply` | 1,000,000 | [3.56, 2.76, 3.47] | [0.727, 0.502, 0.590] |
| `sfm_dlt_points.ply` | 632,970 | [2.41, 2.64, 3.22] | [0.622, 0.604, 0.655] |

`ggpt_points` bbox is 23% tighter than `ff_points` in X-axis (3.56 → 2.72), consistent with expected PTv3 refinement behavior.

vggt_points:  

<img width="884" height="529" alt="Image" src="https://github.com/user-attachments/assets/7e6088f3-e112-4b67-8fd8-9847ceac1bec" />

ggpt_points:

<img width="897" height="523" alt="Image" src="https://github.com/user-attachments/assets/ff63f44b-d085-4123-a4e3-da5a7995056d" />

## Summary

GGPT works on AMD ROCm with minimal changes. The only non-trivial step is the spconv weight layout conversion. Would it be helpful to add a ROCm setup guide or a weight conversion script to the repo?
